api/demo_web_app.py

```diff
@@ -5,31 +5,36 @@
 
 from flask import Flask, request
 
-from .gpt import set_openai_key
+from .gpt import set_openai_key, Example
 from .ui_config import UIConfig
 
 CONFIG_VAR = "OPENAI_CONFIG"
 KEY_NAME = "OPENAI_KEY"
 
+
 def demo_web_app(gpt, config=UIConfig()):
     """Creates Flask app to serve the React app."""
     app = Flask(__name__)
 
     app.config.from_envvar(CONFIG_VAR)
     set_openai_key(app.config[KEY_NAME])
 
-    @app.route("/params", methods=['GET'])
+    @app.route("/params", methods=["GET"])
     def get_params():
         # pylint: disable=unused-variable
         response = config.json()
         return response
 
-    @app.route("/translate", methods=['GET', 'POST'])
+    @app.route("/translate", methods=["GET", "POST"])
     def translate():
         # pylint: disable=unused-variable
-        prompt = request.json['prompt']
+        req_json = request.json
+        prompt = req_json["prompt"]
+        if config.show_example_form and len(req_json["examples"]) > 0:
+            for ex in req_json["examples"]:
+                gpt.add_example(Example(ex["input"], ex["output"]))
         response = gpt.submit_request(prompt)
-        return {'text': response['choices'][0]['text'][7:]}
+        return {"text": response["choices"][0]["text"][7:]}
 
```

Comment:
adding/deleting using UI should maybe use a different route than /translate

Reply:
Hi, thank you for the feedback! I hadn't thought about this!

Instead of having separate endpoints for adding and deleting examples and changing them via AJAX calls, Shall I just clear all the examples and re-add them on every call to /translate? Since the examples are sent again?

The downside is that it makes the /translate endpoint a bit cluttered. But since it's activated only if the show_example_form is True, this should be fine?

What do you think?

Reply:
To me it makes sense to have "Add Example" and "Delete Example" call their own endpoints so that can update just the Examples part of the UI when a change is made. This way we're not re-rendering all the examples and the translation box every time either one is updated. @devstein has more web dev experience, would be good to get his take.

Reply:
An even better way of doing it would be to lazily update the GPT configuartion, that is only write new examples / deleted examples to the GPT object on the back-end when a translate call with those changes is made.

Reply:
I think separate endpoints for add and delete complicates the API. We have to
manage states (indices in this case) across the backend and frontend. One advantage of my
approach is that the frontend is the only source of truth for the examples. 

Also, I think it would show a lag in the UI as you'd have to wait for the `/add`, `/delete` ajax 
requests to complete to actually remove elements (otherwise if a request fails and you 
remove the element before waiting for a response, you can introduce inconsistencies) 
The UI re-renders would be exactly like the current approach + the added lag w.r.t to waiting 
for the AJAX requests.

I agree that it seems wasteful to clear all the examples on each request to `/translate`
but since we would be iterating both the arrays(existing examples and new examples) 
to find the examples that changed and since we're dealing with only a few examples, 
I don't think that approach is computationally helpful.

Let me know what you think and I'll move according to that.

Reply:
I agree with @bora-uyumazturk. I think the examples state should be managed separately from translate. We can achieve this by adding a new endpoints `/examples` that returns all examples on `GET`, creates a new example on `POST`, and deletes an example on a `DELETE` request. This will also remove the difference between Python code examples and UI generated examples because they will all be rendered in the UI regardless of their origin.

To do this, we will need to generate a unique ID from each example. This ID can be used for deleting examples, so we don't have to worry about mapping indices from the Frontend to the Backend. The correct way to do this is to generate a ID by hashing the example's input and output.

```python
import hashlib

class Example():
    """Stores an input, output pair and formats it to prime the model."""

    def __init__(self, inp, out):
        self.input = inp
        self.output = out
        self.id = hashlib.md5(bytes(inp + out, 'utf-8')).hexdigest()
```

Happy to help with the implementation if needed!

Reply:
One doubt I had was regarding how `POST` works in this scenario. Currently the `Add example` button creates a new empty form to enter examples and they're all passed to the `/translate` endpoint. With what you've proposed, would each example form have its own save button? A user adding 3 examples would have to click save thrice. We'd also have to handle edits to examples as well right? Is there a better way to handle this if you still want to have a separate endpoint for examples?

Reply:
Great point, in this case, we do not want to generate the Example ID by hashing the input and output, but rather a random [uuid](https://docs.python.org/3.8/library/uuid.html). After creating it, you can save (`PUT`) the example while the user types. Just make sure to [debounce](https://lodash.com/docs/4.17.14#debounce) the save (`PUT`) requests

Reply:
I have added the required changes. Let me know if there is anything missing.